### PR TITLE
Go for C# people part one

### PR DESCRIPTION
**Pull request by kfcampbell**
This is a blog post that introduces Go for developers with C# experience.

If you're reading this, please read the post and maybe offer some criticism!

Or don't. I'm a PR, not a cop.

👀 [Rendered version](https://kfcampbell-85a55c161a.drafts.github.io/blog/2020/go-for-csharp-people-part-one/) 👀

### CODE REVIEW

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+Before we can get into modules, however, we need to briefly discuss packages. All Go source files must declare a package in their first line, and every Go file must have at least one `package main` file that declares a `func main()`. Other than the file that contains `main`, code must be placed in a directory whose name matches the `package` of files within it. As previously mentioned, all uncapitalized members and functions are private, and all capitalized ones are publically visible. 
+
+Modules allow reusable packages to be imported using their URL, as in `import github.com/go-sql-driver/mysql`. Interestingly, packages from within a project as well as third party dependencies are imported using _the exact same syntax_. Put another way: if a function is public in your code, and your code is accessible, anybody can import and use it! This can lead to some weird issues [like this one](https://github.com/github/hub/issues/2517#issuecomment-615531677), in which parts of an application are consumed as a library by other applications.
```
Comment (davecheney) [inline _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: yeah, its worthwhile digging into this as, compared to C#, the access and protection modifiers on types (I'm thinking -- sorry, java speak -- default, protect, private, protected), go only has "exported to the world" and "pacakge private", which is probably a sigifnicant point of difference for -- i'm guessing -- nuget's explicit export declaration.

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+## Quick Go Summary
+
+Go was developed at Google, first announced in 2009 and initially released in 2012. The language is:
+
+- statically-typed
+- compiled
+- automatically garbage-collected
+
+Go programs compile down to a single binary with all dependencies included. In contrast to C#, no runtime is needed, so applications can be run in very thin containers such as [alpine](https://hub.docker.com/_/alpine/).
```
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: Good point. Added!

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+In addition, Go has a strong focus on concurrency. The approach it takes is a version of [Communicating Sequential Processes](https://en.wikipedia.org/wiki/Communicating_sequential_processes), which is a formal way of specifying rules for interaction in concurrent systems whose details are beyond the scope of this post. For our purposes, it's enough to know that Go "shares memory by communicating", in which various concurrent processes called "goroutines" pass objects between them. This is a large departure from more traditional object-oriented concurrency schemes, in which applications instead "communicate by sharing memory": traditionally multiple threads share a reference to an object, and writes are controlled through a delicate balance of locks and mutexes.
+
+Go is frequently compared to C due to a similar syntax, pointer usage, and relatively un-object-oriented nature. However, Go is much safer than C (it's much harder to cut yourself on a sharp pointer) and its concurrency tooling and robust standard library means it offers significant capability on top of vanilla C.
```
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: Good idea. Changed!

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+## Goals of this Blog Post
+
+Welcome! This post assumes working knowledge of C# and it's targetted toward those without Go experience. The aim is to provide just enough of an introduction to the language to make you dangerous :P.  When I start learning a new language, I tend to learn best by doing - that is, I like to experiment and make the dumb mistakes myself, before the deeper lessons sink in, and I'd like to make this the sort of blog post I wish I had when I was learning Go for the first time.
+
+That being said, there's a couple basic things we have to discuss up front before we can jump straight into writing Go code (I know, I'm sorry!). But trust me, it won't be so bad, and we'll let you stare at compiler errors yourself in Part Two (coming soon!). For now, let's start with a brief overview of the language and some of its major design decisions.
+
+## Quick Go Summary
+
+Go was developed at Google, first announced in 2009 and initially released in 2012. The language is:
```
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown:28]: Gotta love that SEO! I'll throw these links in.

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+### But Different
+
+- Go has no runtime required (like the [.NET CLR](https://docs.microsoft.com/en-us/dotnet/standard/clr))
+- Go has no generics
+    - This is a consistent paint point for the Go community
+    - See [here](https://go.googlesource.com/proposal/+/master/design/go2draft-generics-overview.md) for a proposal to add generics from language designer Russ Cox
+- Go returns errors which are then checked by the caller, rather than try/catch blocks and exceptions for error handling
+- Go has no concept of `await`ing an action (more on this later)
+- Go places a strong emphasis on staying up to date with the language, in constrast to the fragmentation in the .NET and .NET Core ecosystems
```
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown:86]: Hmm...I'm not sure. Let me look into it!

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+## Dependency management
+
+Dependency management used to be a somewhat contentious subject in the Go ecosystem, with a number of third party solutions attempting to address the gap in this language tooling. [v1.11](https://golang.org/doc/go1.11), however, introduced experimental support for Go modules, and [v1.14](https://golang.org/doc/go1.14) (released this February) standardized them as ready for production use.
+
+Before we can get into modules, however, we need to briefly discuss packages. All Go source files must declare a package in their first line, and every Go file must have at least one `package main` file that declares a `func main()`. Other than the file that contains `main`, code must be placed in a directory whose name matches the `package` of files within it. As previously mentioned, all uncapitalized members and functions are private, and all capitalized ones are publically visible.
```
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: That's an excellent idea. I swapped things around a little bit, and I think it reads better now.

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown (hunk 1 above)
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: I've elaborated on this a bit, and I think it's a bit better now. Thanks!

File: _posts/2020-05-21-go-for-csharp-people-part-one.markdown (hunk 5 above)
[the remark this answers is not in the document]
Comment (kfcampbell) [reply _posts/2020-05-21-go-for-csharp-people-part-one.markdown]: Updated with a comment about both languages' backwards compatibility. Thanks for bringing this up!

**Pull request merged**